rpm/0070-sailfishos-webrtc-Implement-video-capture-module.-JB.patch

```diff
@@ -0,0 +1,2255 @@
+index 000000000000..fd3c8bd35632
+--- /dev/null
++++ b/media/webrtc/trunk/webrtc/modules/video_capture/sfos/video_capture_sfos.cc
+@@ -0,0 +1,174 @@
++/*
++ *  Copyright (c) 2021 Open Mobile Platform LLC. All Rights Reserved.
++ *
++ *  Use of this source code is governed by a BSD-style license
++ *  that can be found in the LICENSE file in the root of the source
++ *  tree. An additional intellectual property rights grant can be found
++ *  in the file PATENTS.  All contributing project authors may
++ *  be found in the AUTHORS file in the root of the source tree.
++ */
++
++#include <errno.h>
++#include <fcntl.h>
++#include <stdio.h>
++#include <string.h>
++#include <sys/ioctl.h>
++#include <sys/mman.h>
++#include <sys/stat.h>
++#include <unistd.h>
++
++#include <new>
++
++#include "webrtc/system_wrappers/include/trace.h"
++#include "webrtc/system_wrappers/include/clock.h"
++
++#include "webrtc/modules/video_capture/sfos/video_capture_sfos.h"
++
++namespace webrtc {
++namespace videocapturemodule {
++rtc::scoped_refptr<VideoCaptureModule> VideoCaptureImpl::Create(
++    const char* deviceUniqueId) {
++    rtc::scoped_refptr<VideoCaptureModuleSFOS> implementation(
++        new rtc::RefCountedObject<VideoCaptureModuleSFOS>());
++
++    if (implementation->Init(deviceUniqueId) != 0)
++        return nullptr;
++
++    return implementation;
++}
++
++VideoCaptureModuleSFOS::VideoCaptureModuleSFOS()
++    : VideoCaptureImpl()
++{
++    mozilla::hal::ScreenConfiguration screenConfig;
++
++    mozilla::hal::RegisterScreenConfigurationObserver (this);
++    mozilla::hal::GetCurrentScreenConfiguration (&screenConfig);
++    _screenRotationAngle = ScreenOrientationToAngle (screenConfig.orientation());
++}
++
++int32_t VideoCaptureModuleSFOS::Init(const char* deviceUniqueIdUTF8)
++{
++    int len = strlen(deviceUniqueIdUTF8);
++
++    /* Fill the current device name */
++    _deviceUniqueId = new (std::nothrow) char[len + 1];
++    if (_deviceUniqueId) {
++        memcpy(_deviceUniqueId, deviceUniqueIdUTF8, len + 1);
++    } else {
++        return -1;
++    }
++
++    auto cameraManager = GetCameraManager(SFOS_CAMERA_TYPE_GST_DROID);
++    if (cameraManager) {
++        _camera = cameraManager->OpenCamera(deviceUniqueIdUTF8);
++        if (_camera) {
++            _camera->SetListener(this);
++            return 0;
++        }
++    }
++
++    return -1;
++}
++
++VideoCaptureModuleSFOS::~VideoCaptureModuleSFOS()
++{
++    mozilla::hal::UnregisterScreenConfigurationObserver(this);
++}
++
++int32_t VideoCaptureModuleSFOS::StartCapture(
++    const VideoCaptureCapability& capability)
++{
++    _requestedCapability = capability;
++    _startNtpTimeMs = webrtc::Clock::GetRealTimeClock()->CurrentNtpInMilliseconds();
++    return _camera->StartCapture(capability);
++}
++
++int32_t VideoCaptureModuleSFOS::StopCapture()
++{
++    return _camera->StopCapture();
++}
++
++bool VideoCaptureModuleSFOS::CaptureStarted()
++{
++    return _camera->CaptureStarted();
```

Comment:
Is it safe to assume `_camera != null` for these?

Reply:
Yes. The module will not be created it it has not passed `Init()` in `VideoCaptureImpl::Create()`.

Reply:
Thanks for clarifying.